### PR TITLE
Fixes  #3713 - `TextField` & `TextView` - When pressing right/left clear the selection instead of move tab

### LINKED ISSUE #3713: Arrow Right in TextField/TextView should not move tab if there is a selection
**Describe the bug**
1. make a selection up to the end of the text by holding shift and pressing right
2. release shift
3. press right

Expected behaviour: selection should clear and cursor should be at end, the same as if you had pressed the End key.  Tab should only occur on second press of Right i.e. when there is no longer a selection.

Observed behaviour: tab moves to next view

Happy to discuss if people disagree

image
_TextField has an active selection so right cursor could be resolved by clearing that selection_

### PR DESCRIPTION
**Pull request by tznind**
## Fixes

- Fixes #3713

## Pull Request checklist:

- [x] I've named my PR in the form of "Fixes #issue. Terse description."
- [x] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [x] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [x] I conducted basic QA to assure all features are working

### CODE REVIEW

File: Terminal.Gui/Views/TextField.cs [hunk 1]
```diff
@@ -1531,19 +1531,27 @@ private void MoveHomeExtend ()
...
-            return true;
-        }
+        _cursorPosition = Math.Min (_text.Count, Math.Max (0, _cursorPosition + distance));
+        ClearAllSelection ();
+        Adjust ();
 
-        return false;
+        return _cursorPosition != oldCursorPosition || hadSelection;
+    }
+    private bool MoveLeft ()
```
Comment (BDisp) [inline Terminal.Gui/Views/TextField.cs:1552]: Please leave a space between them.

File: UnitTests/Views/TextFieldTests.cs [hunk 2]
```diff
@@ -2027,6 +2027,27 @@ public void Autocomplete__Added_To_SuperView_On_Add ()
         Assert.Equal (2, superView.Subviews.Count);
     }
 
+    [Fact]
+    public void Right_CursorAtEnd_WithSelection_ShouldClearSelection ()
```
Comment (BDisp) [inline UnitTests/Views/TextFieldTests.cs:2031]: Please don't forget to add a unit test for Left:

```cs
    [Fact]
    public void Left_CursorAtStart_WithSelection_ShouldClearSelection ()
```

**Review by BDisp — CHANGES_REQUESTED** [review]
LGTM. Some minor things. Thanks

File: UnitTests/Views/TextFieldTests.cs [hunk 3]
```diff
@@ -2048,7 +2048,32 @@ public void Right_CursorAtEnd_WithSelection_ShouldClearSelection ()
...
+        {
+            Text = "Hello",
+        };
+        tf.SetFocus ();
+
+        tf.CursorPosition = 2;
+        Assert.True (tf.NewKeyDownEvent (Key.CursorLeft.WithShift));
+        Assert.True (tf.NewKeyDownEvent (Key.CursorLeft.WithShift));
 
+        // When there is selected text and the cursor is at the end of the text field
```
Comment (BDisp) [inline UnitTests/Views/TextFieldTests.cs:2064]: Replace end to start.

**Review by BDisp — CHANGES_REQUESTED** [review]

File: UnitTests/Views/TextFieldTests.cs (hunk 3 above)
Comment (tznind) [reply UnitTests/Views/TextFieldTests.cs]: good spot thanks

File: UnitTests/Views/TextFieldTests.cs (hunk 2 above)
Comment (tznind) [reply UnitTests/Views/TextFieldTests.cs:2031]: have done

File: UnitTests/Views/TextViewTests.cs [hunk 4]
```diff
@@ -8660,4 +8660,58 @@ public void Autocomplete_Visible_False_By_Default ()
...
+
+        tv.NewKeyDownEvent (Key.CursorRight);
+        tv.NewKeyDownEvent (Key.CursorRight);
+
+        Assert.Equal (2,tv.CursorPosition.X);
+
+        Assert.True (tv.NewKeyDownEvent (Key.CursorLeft.WithShift));
+        Assert.True (tv.NewKeyDownEvent (Key.CursorLeft.WithShift));
+
+        // When there is selected text and the cursor is at the end of the text field
```
Comment (BDisp) [inline UnitTests/Views/TextViewTests.cs:8704]: Replace end to start. Copy/paste 😄

**Review by BDisp — CHANGES_REQUESTED** [review]

**Review by BDisp — APPROVED** [review]

**Review by tig — APPROVED** [review]

**Comment by tig** [comment]
@tznind please fix the conflicts.

**Comment by tznind** [comment]
> @tznind please fix the conflicts.

Have fixed

**Pull request merged**